packages/plugins/rubygem-smart_proxy_salt/rubygem-smart_proxy_salt.spec

```
@@ -177,18 +179,25 @@ EOF
 %{foreman_proxy_bundlerd_dir}/%{plugin_name}.rb
 %exclude %{gem_cache}
 %{gem_spec}
-%{foreman_proxy_bundlerd_dir}/smart_proxy_salt_core.rb
-%{smart_proxy_dynflow_bundlerd_dir}/smart_proxy_salt_core.rb
 %config(noreplace) %{salt_config_dir}/foreman.yaml
 %config %{_root_sysconfdir}/cron.d/%{gem_name}
 %exclude %{gem_instdir}/etc
 %exclude %{gem_instdir}/cron
+%attr(-,foreman-proxy,foreman-proxy) %{salt_proxy_runners_dir}/foreman_file.py*
+%attr(-,foreman-proxy,foreman-proxy) %{salt_proxy_runners_dir}/foreman_https.py*
+%attr(-,foreman-proxy,foreman-proxy) %{salt_proxy_reactors_dir}/foreman_minion_auth.sls
+%attr(-,foreman-proxy,foreman-proxy) %{salt_state_grains_dir}/autosign_key
```

does this file need to be writable by the proxy, because it's used as a mechanism to share data between `smart_proxy_salt` and the salt master?

---

Exactly, `smart_proxy_salt` adds entries to the file and salt master reads the file during minion authentication.

---

I fear that as it is now a `yum reinstall` or `yum update` will clear this file back to its packaged state. Wouldn't it be better to have it create the directory (with the correct permissions) and expect the Smart Proxy to create the file at runtime?
```suggestion
%dir %attr(-,foreman-proxy,foreman-proxy) %{salt_state_grains_dir}
```

You may also need to add the `autosign_key` file as a `%ghost` to ensure it's properly owned.

I'm also looking at @evgeni here for advice on what's best, but I think the replacement on upgrades is a real issue.

---

oh right, this is outside of `/etc`, so RPM will replace it, good catch!
Yeah, I think just creating the dir and `%ghost`ing the file is the correct approach here from a packaging perspective.

But! ;-)

the proxy plugin explicitly doesn't create this file, but expects it to be present: https://github.com/theforeman/smart_proxy_salt/blob/ddaef0df69c850015a6ec98e425dd58b9339c760/lib/smart_proxy_salt/cli.rb#L67

I don't see a good reason for it not to create it, but maybe I don't see the whole picture with salt…

---

I'm not sure why it fails on that: `open(filepath, File::RDWR)` should be able to create it, right?

@bastian-src Some unrelated (to this PR) comments that I saw while reading through that file. I'd recommend this pattern to guarantee the file descriptor is closed:
```ruby
open(filepath, File::RDWR) do |file|
  found = false
  file.each_line { |line| found = true if line.chomp == value }
  file.puts value unless found
end
```

I also like to use `File.open` instead of plain `open`. This can avoid accidentally exposing much more. For example, `open()` can also be used to open a subshell (by starting it with `|`). That can escalate a vulnerability from writing to arbitrary files to a remote code execution.

Also, in https://github.com/theforeman/smart_proxy_salt/blob/ddaef0df69c850015a6ec98e425dd58b9339c760/lib/smart_proxy_salt/cli.rb#L89 you're leaking a file descriptor since it's never closed. You can use `File.read()` as well if you just want to read the file completely.

And I think https://github.com/theforeman/smart_proxy_salt/blob/ddaef0df69c850015a6ec98e425dd58b9339c760/lib/smart_proxy_salt/cli.rb#L101-L103 could be replaced with `File.write(filepath, entries.join())`.

---

> I don't see a good reason for it not to create it, but maybe I don't see the whole picture with salt…

The `autosign_key` file contains an authentication key during the host creation which is used to authenticate the Salt Minion against the Salt Master initially. Afterwards, the key is removed from the file.
Accordingly, we would only run into problems if a package update/reinstall is performed during the deployment of a host. Then, the Salt Minion wouldn't be authenticated automatically and would've to be accepted manually.

I don't think this is a scenario we should actually care about too much/support at all.
Still, do you think it would be better to let Smart Proxy Salt create the `autosign_key` file instead of foreman-packaging?

@ekohl thanks for the advice! I'll have a look at the file.

---

I think it would be better if Smart Proxy Salt could create it. The major benefit is that a user can remove the file and it can recover from it. Right now it will just "crash" requiring manual admin intervention.

And I do think it's important to deal with this. It would feel unreliable to the user. Losing data is always bad and it makes the user question what else it will just throw away.

---

Alright, I agree that it enhances the user experience in case the file doesn't exist for some reason.

I've created [this](https://github.com/theforeman/smart_proxy_salt/pull/74) PR and tested it right away.
Unfortunately, I'm getting a `Permission denied @ rb_sysop` error. Could this be related to SELinux?